File: doubly_linked_list/doubly_linked_list.c
#include<stdio.h>
#include<stdlib.h>

//create linklist node  structure using struct
struct linkListNode{
	struct linkListNode *previous;
	int data;
	struct linkListNode *next;
};
typedef struct linkListNode *node;

node head=NULL;
node start=NULL;

addNode(int);
deleteNode(int);
node createNode();

//code for create node
node createNode(){
	node newNode;
	newNode=(node)malloc(sizeof(node));
	return newNode;
}

//code for add a node into the list
addNode(int data){
	node myNode=createNode();
	myNode->previous=head;
	myNode->data=data;
	myNode->next=NULL;
	if(head==NULL)
		start=myNode;
	else
		head->next=myNode;
	head=myNode;
}

//code for delete a given node
//it may be 1st or last or any
deleteNode(int data){
	node temp=start;
	while(temp->data!=data){
		temp=temp->next;
	}
	if(temp==start){
		start=start->next;	
		start->previous=NULL;
	}else if(temp->next==NULL){
		temp->previous->next=NULL;
	}else{
		temp->previous->next=temp->next;
		temp->next->previous=temp->previous;
	}
	free(temp);
} 
int main(){
	//call hear above function..
	//addNode(int) this function for create and add a node to linklist..
	//deleteNode(int) this function for delate any given node...
	return 0;
}

File: data-structures/doubly-linked-list/doubly_linked_list.py
class Node:    
    def __init__(self,data):    
        self.data = data;    
        self.previous = None;    
        self.next = None;    
            
class DoublyLinkedList:    
    def __init__(self):    
        self.head = None;    
        self.tail = None;    
            
  
    def addNode(self, data):
        newNode = Node(data);    
                
        if(self.head == None):    
            self.head = self.tail = newNode;    
            self.head.previous = None;    
            self.tail.next = None;    
        else:    
            self.tail.next = newNode;    
            newNode.previous = self.tail;    
            self.tail = newNode;    
            self.tail.next = None;    
                
    def display(self):    
        current = self.head;    
        if(self.head == None):    
            print("Doubly Link List is empty");    
            return;    
        print("Nodes");    
        while(current != None):     
            print(current.data),;    
            current = current.next;    
                
dList = DoublyLinkedList();    


dList.addNode(54);    
dList.addNode(27);    
dList.addNode(76);    
dList.addNode(65);    
dList.addNode(4);    
     

dList.display();  
